File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int max_profit = 0;
        int buy_price = prices[0];

        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < buy_price) {
                // finding lowest price to buy stock on
                buy_price = prices[i];
            } else if (prices[i] - buy_price > max_profit) {
                // attempts to find highest profit
                // prices[i] is curr price to sell stock on
                max_profit = prices[i] - buy_price;
            }
        }

        return max_profit;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        minBuy = prices[0]

        # sliding window
        for i in range(1, len(prices)):
            minBuy = min(minBuy, prices[i])
            res = max(res, prices[i] - minBuy)

        return res